examples/Calculator.re

```diff
@@ -5,245 +5,341 @@ open Revery.UI.Components;
-            (display ++ curNum ++ newOpString, result /. float_of_string(curNum));
-          } else {
-            ("Error: Divide by zero!", result);
-          }
-        };
-        setResult(newResult);
-        setDisplay(newDisplay);
-        newResult;
-      };
-
-      let perform(newOp) {
-        let _newResult = eval(newOp);
-        setOp(newOp);
-        setCurNum("");
-      };
-
-      let calculate() {
-        let newResult = eval(`Nop);
-        setOp(`Nop);
-        setCurNum(showFloat(newResult));
-      };
+type action =
+  | BackspaceKeyPressed
+  | ClearKeyPressed(bool) /*AC pressed */
+  | DotKeyPressed
+  | NumberKeyPressed(string)
+  | OperationKeyPressed(operator)
+  | PlusMinusKeyPressed
+  | ResultKeyPressed;
 
-      let clear(~clearAcc) {
-        if (clearAcc) {
-          setOp(`Nop);
-          setResult(0.);
-          setDisplay("");
-        };
-        setCurNum("");
-      };
+let eval = (state, newOp) => {
+  /* Figure out what the string for the next operation will be */
+  let newOpString =
+    switch (newOp) {
+    | `Nop => ""
+    | `Add => "+"
+    | `Sub => "-"
+    | `Mul => "×"
+    | `Div => "÷"
+    };
+  /* Split the current display on ! and get everything after (to clear errors) */
+  let partitionedDisplay = String.split_on_char('!', state.display);
+  let display =
+    List.nth(partitionedDisplay, List.length(partitionedDisplay) - 1);
+  let (newDisplay, newResult) =
+    switch (state.operator) {
+    | #operator when state.number == "" => (
+        "Error: Can't evaluate binary operator without input!",
+        state.result,
+      )
+    | `Nop => (state.number ++ newOpString, float_of_string(state.number))
+    | `Add => (
+        display ++ state.number ++ newOpString,
+        state.result +. float_of_string(state.number),
+      )
+    | `Sub => (
+        display ++ state.number ++ newOpString,
+        state.result -. float_of_string(state.number),
+      )
+    | `Mul => (
+        display ++ state.number ++ newOpString,
+        state.result *. float_of_string(state.number),
+      )
+    | `Div =>
+      if (float_of_string(state.number) != 0.) {
+        (
+          display ++ state.number ++ newOpString,
+          state.result /. float_of_string(state.number),
+        );
+      } else {
+        ("Error: Divide by zero!", state.result);
+      }
+    };
+  (newResult, newDisplay);
+};
 
-      let negate() {
-        if (curNum != "" && curNum.[0] == '-') {
-          setCurNum(String.sub(curNum, 1, String.length(curNum) - 1));
-        } else {
-          setCurNum("-" ++ curNum);
-        };
+let reducer = (state, action) =>
+  switch (action) {
+  | BackspaceKeyPressed =>
+    state.number == "" ?
+      state :
+      {
+        ...state,
+        number: String.sub(state.number, 0, String.length(state.number) - 1),
+      }
+  | ClearKeyPressed(ac) =>
+    ac ?
+      {...state, operator: `Nop, result: 0., display: ""} :
```

Comment:
Shouldn't this also clear the `number` field? Otherwise, this'll only clear the current equation but leave the current value "typed in"

Reply:
Oh right, great catch! I thought the existing statement was an if/else but the reset of `curNum` was always happening. Fixed in https://github.com/OhadRau/revery/pull/1/commits/240b993fb18015795788506d4c0d4ec83a298475. 👍